### PR TITLE
Feature: Allowed domains for chatflows

### PR DESCRIPTION
**Pull request by 0xi4o**
- Adds an allowed domains dialog for chatflows where users can whitelist which domains can access the chatflow. By default all domains are allowed. If a list is present, then only those domains can use the chatflow prediction, either from embed or API.

flowise-allowed-domains-settings
flowise-allowed-domains-settings-2
Flowise-Test

### CODE REVIEW

**Comment by automaton82** [comment]
I'm a little confused why you'd use this over CORS or CSP? They do the same thing but at a lower level to ensure no access can be done at all except from the whitelisted domains.

**Comment by HenryHengZJ** [comment]
the intention is to allow the chatbot to be embedded on specific domain:
image

**Comment by automaton82** [comment]
Yea I get that but that's the same thing CORS does. Except CORS will pass security tests since the preflight options will also fail when not on an allowed domain.

It's the same functionality just at the level where it's expected. Otherwise they'll be tempted to set CORS to * and use this new config instead which isn't secure, since any domain will be able to access the REST API, regardless of this setting.

If you remember I mentioned this on the other ticket. The very first request xdomain sends is OPTIONS preflight check. This is before any logic can be done about a chatbot ID, thus CORS is in the HTTP pipeline and configured at the env level.

**Comment by automaton82** [comment]
If your goal is to solve multi tenancy in cloud hosting (1 host per N clients) then better to solve it with actual accounts and tenancy, where you keep the tenancy as the initial request and sticky it somehow through all subsequent requests.

Anyways just my $0.02 😄

**Comment by HenryHengZJ** [comment]
> Yea I get that but that's the same thing CORS does. Except CORS will pass security tests since the preflight options will also fail when not on an allowed domain.
> 
> It's the same functionality just at the level where it's expected. Otherwise they'll be tempted to set CORS to * and use this new config instead which isn't secure, since any domain will be able to access the REST API, regardless of this setting.
> 
> If you remember I mentioned this on the other ticket. The very first request xdomain sends is OPTIONS preflight check. This is before any logic can be done about a chatbot ID, thus CORS is in the HTTP pipeline and configured at the env level.

yeah I think there are 2 level of users for Flowise, for devs its always welcomed to set the env variables CORS for that. For no-code users they might not be familiar with the CORS and setting env variables, so this is an alternative, although it doesnt solve at the very root level, but still offers some protective mechanism at the very least

**Comment by 0xi4o** [comment]
@HenryHengZJ I've updated the error message for a non-whitelisted domain. Please review this.

File: packages/server/src/index.ts [hunk 1]
```diff
@@ -1319,7 +1319,32 @@ export class App {
...
+                const chatflow = await this.AppDataSource.getRepository(ChatFlow).findOneBy({
+                    id: req.params.id
+                })
+                if (!chatflow) return res.status(404).send(`Chatflow ${req.params.id} not found`)
+                let isDomainAllowed = true
+                logger.info(`[server]: Request originated from ${req.headers.origin}`)
+                if (chatflow.chatbotConfig) {
+                    const parsedConfig = JSON.parse(chatflow.chatbotConfig)
+                    // check whether the first one is not empty. if it is empty that means the user set a value and then removed it.
+                    const isValidAllowedOrigins = parsedConfig.allowedOrigins[0] !== ''
```
[the remark this answers is not in the document]
Comment (HenryHengZJ) [reply packages/server/src/index.ts:1331]: yep, it does error:
image

File: packages/server/src/index.ts [hunk 2]
```diff
@@ -1319,7 +1319,32 @@ export class App {
...
+                    if (parsedConfig.allowedOrigins && parsedConfig.allowedOrigins.length > 0 && isValidAllowedOrigins) {
+                        const originHeader = req.headers.origin as string
+                        const origin = new URL(originHeader).host
+                        isDomainAllowed =
+                            parsedConfig.allowedOrigins.filter((domain: string) => {
+                                const allowedOrigin = new URL(domain).host
```
Comment (HenryHengZJ) [inline packages/server/src/index.ts:1337]: if `domain` is invalid, you will get error `Invalid URL` and the apps shut down due to unhandled rejection

**Review by HenryHengZJ — APPROVED** [review]

**Pull request merged**